src/lib/storage/frame_of_reference_segment/frame_of_reference_segment_iterable.hpp

```diff
@@ -36,13 +36,12 @@ class FrameOfReferenceSegmentIterable : public PointAccessibleSegmentIterable<Fr
   template <typename Functor>
   void _on_with_iterators(const std::shared_ptr<const PosList>& position_filter, const Functor& functor) const {
     _segment.access_counter[SegmentAccessCounter::access_type(*position_filter)] += position_filter->size();
-    resolve_compressed_vector_type(_segment.offset_values(), [&](const auto& vector) {
-      auto decompressor = vector.create_decompressor();
-      using OffsetValueDecompressorT = std::decay_t<decltype(decompressor)>;
+    resolve_compressed_vector_type(_segment.offset_values(), [&](const auto& offset_values) {
+      using OffsetValueDecompressorT = std::decay_t<decltype(offset_values.create_decompressor())>;
 
       auto begin = PointAccessIterator<OffsetValueDecompressorT>{&_segment.block_minima(), &_segment.null_values(),
-                                                                 std::move(decompressor), position_filter->cbegin(),
-                                                                 position_filter->cbegin()};
+                                                                 offset_values.create_decompressor(),
+                                                                 position_filter->cbegin(), position_filter->cbegin()};
 
       auto end = PointAccessIterator<OffsetValueDecompressorT>{position_filter->cbegin(), position_filter->cend()};
```

Comment:
This made it through the review back then, but is actually pretty awful. There should be no special constructor for the "end iterator", because there is no such thing. There is an iterator pointing to the end. Decrementing and dereferencing that iterator should be possible. Here is a diff for a test:

```
diff --git a/src/test/storage/iterables_test.cpp b/src/test/storage/iterables_test.cpp
index 407a7ae3..a45fe51d 100644
--- a/src/test/storage/iterables_test.cpp
+++ b/src/test/storage/iterables_test.cpp
@@ -167,11 +167,42 @@ TEST_P(EncodedSegmentIterablesTest, IteratorWithIterators) {
     using SegmentType = std::decay_t<decltype(segment)>;

     if constexpr (!std::is_same_v<pmr_string, ColumnDataType>) {
-      auto sum = ColumnDataType{0};
-      auto accessed_offsets = std::vector<ChunkOffset>{};
-      const auto functor = SumUpWithIterator<ColumnDataType>{sum, accessed_offsets};
+      // Test summing up all values in the segment or all values referenced by the PosList
+      {
+        auto sum = ColumnDataType{0};
+        auto accessed_offsets = std::vector<ChunkOffset>{};
+        const auto functor = SumUpWithIterator<ColumnDataType>{sum, accessed_offsets};
+
+        const auto iterable = create_iterable_from_segment<ColumnDataType, false /* no type erasure */>(segment);
+        if (with_position_filter) {
+          if constexpr (!std::is_same_v<SegmentType, ReferenceSegment>) {
+            iterable.with_iterators(position_filter, functor);
+          }
+        } else {
+          iterable.with_iterators(functor);
+        }

+        EXPECT_EQ(sum, expected_sum(nullable, with_position_filter));
+        EXPECT_EQ(accessed_offsets, expected_offsets(with_position_filter));
+      }
+    }
+
+    // Next, test that begin and end iterators are compatible (i.e., that iterating from both ends allows us to meet in
+    // the middle)
+    {
       const auto iterable = create_iterable_from_segment<ColumnDataType, false /* no type erasure */>(segment);
+      const auto functor = [&](auto begin, auto end) {
+        if (with_position_filter) {
+          begin += position_filter->size() - 1;
+        } else {
+          begin += base_segment->size() - 1;
+        }
+        --end;
+
+        EXPECT_EQ(begin->value(), end->value());
+        EXPECT_EQ(begin->chunk_offset(), end->chunk_offset());
+      };
+
       if (with_position_filter) {
         if constexpr (!std::is_same_v<SegmentType, ReferenceSegment>) {
           iterable.with_iterators(position_filter, functor);
@@ -179,9 +210,6 @@ TEST_P(EncodedSegmentIterablesTest, IteratorWithIterators) {
       } else {
         iterable.with_iterators(functor);
       }
-
-      EXPECT_EQ(sum, expected_sum(nullable, with_position_filter));
-      EXPECT_EQ(accessed_offsets, expected_offsets(with_position_filter));
     }
   });
 }
```

Currently, this segfaults.

Reply:
> So for the containers listed, not only should the iterator returned from end() be decrementable, the decremented iterator should also be dereferencable. (Unless the container is empty, of course. That invokes undefined behavior.)

https://stackoverflow.com/questions/5322104/how-portable-is-end-iterator-decrement

Reply:
I removed this end iterator now as well. Did not think too much about it before, but yes. It's quite bad. Sorry for that overlook.
When I tried reference wrapper, it directly forbid to use nullptrs which I would see as an advantage of std::reference_wrapper (and there appears to be no runtime costs).